src/main/java/io/strimzi/kafka/bridge/tracing/OpenTelemetryHandle.java

```diff
@@ -0,0 +1,259 @@
+                spanBuilder.addLink(psc);
+            }
+        }
+        spanBuilder
+            .setSpanKind(SpanKind.CONSUMER)
+            .setParent(Context.current())
+            .startSpan()
+            .end();
+    }
+
+    private static TextMapPropagator propagator() {
+        return GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+    }
+
+    private static final TextMapGetter<RoutingContext> RCG = new TextMapGetter<RoutingContext>() {
+        @Override
+        public Iterable<String> keys(RoutingContext rc) {
+            return rc.request().headers().names();
+        }
+
+        @Override
+        public String get(RoutingContext rc, String key) {
+            if (rc == null) {
+                return null;
+            }
+            return rc.request().headers().get(key);
+        }
+    };
+
+    private static final TextMapGetter<Map<String, String>> MG = new TextMapGetter<Map<String, String>>() {
+        @Override
+        public Iterable<String> keys(Map<String, String> map) {
+            return map.keySet();
+        }
+
+        @Override
+        public String get(Map<String, String> map, String key) {
+            return map != null ? map.get(key) : null;
+        }
+    };
+
+    @Override
+    public <K, V> SpanHandle<K, V> span(RoutingContext routingContext, String operationName) {
+        return buildSpan(getSpanBuilder(routingContext, operationName), routingContext);
+    }
+
+    private static <K, V> SpanHandle<K, V> buildSpan(SpanBuilder spanBuilder, RoutingContext routingContext) {
+        spanBuilder.setSpanKind(SpanKind.SERVER);
+        setCommonAttributes(spanBuilder, routingContext);
+        return new OTelSpanHandle<>(spanBuilder.startSpan());
+    }
+
+    private static class OTelSpanBuilderHandle<K, V> implements SpanBuilderHandle<K, V> {
+        private final SpanBuilder spanBuilder;
+
+        public OTelSpanBuilderHandle(SpanBuilder spanBuilder) {
+            this.spanBuilder = spanBuilder;
+        }
+
+        @Override
+        public SpanHandle<K, V> span(RoutingContext routingContext) {
+            return buildSpan(spanBuilder, routingContext);
+        }
+    }
+
+    @Override
+    public void kafkaConsumerConfig(Properties props) {
+    }
+
+    @Override
+    public void kafkaProducerConfig(Properties props) {
+        props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, ContextAwareTracingProducerInterceptor.class.getName());
+    }
+
+    private static final class OTelSpanHandle<K, V> implements SpanHandle<K, V> {
+        private final Span span;
+        private final Scope scope;
+
+        public OTelSpanHandle(Span span) {
+            this.span = span;
+            this.scope = span.makeCurrent();
+        }
+
+        @Override
+        public void prepare(KafkaProducerRecord<K, V> record) {
+            String uuid = UUID.randomUUID().toString();
+            SPANS.put(uuid, span);
+            record.addHeader(X_UUID, uuid);
```

Comment:
I tried to follow the flow. I see that when we send a record, you are calling this `prepare` to generate a UUID and it's put in the header of the Kafka record but when the `onSend` of the interceptor is called, you are removing this UUID before sending the record. So what's the purpose of this UUID in the record header between the prepare and onSend?

Reply:
My question was not related to the `SPANS` map but to the headers in the Kafka record. You are adding the UUID in the header of the record here, but then before sending the record (in the interceptor) you are removing the UUID ... so I don't understand the reason for having the UUID being added here (then removed, without any use, later).

Reply:
Actually looking at the code even OpenTracing sets a `TracingProducerInterceptor` class in the configuration but there is no special implementation of it compared to OpenTelemetry (your `ContextAwareTracingProducerInterceptor`). Why?

Reply:
I think I have got the reason. You are getting the HTTP bridge "send" span and you are propagate that span as parent on the interceptor send thread so that it will be parent of the topic send provided by the base `TracingProducerInterceptor` implementation. My only question is why all this stuff is not needed in OpenTracing ...

Reply:
> My only question is why all this stuff is not needed in OpenTracing

It looks like OpenTracing propagates current span differently - I guess not via ThreadLocal ...
I didn't go into details for OpenTracing, since it worked as expected.

This surprised me in OTel, and was actually tricky to find this async send at first - this resulted in spans not being related, hence the "hacky" interceptor workaround.

Reply:
This is where OpenTracing works. The `onSend` of its default interceptor calls this `buildAndInjectSpan` which is able to extract the current span context and set it as child.

https://github.com/opentracing-contrib/java-kafka-client/blob/31ce5260279c2cadf3d69c0acbd50f024afe4660/opentracing-kafka-client/src/main/java/io/opentracing/contrib/kafka/TracingKafkaUtils.java#L80

So following the `TracingKafkaUtils.extraSpanContext`, it's doing some action through the record headers where the context is stored and is able to get it.

https://github.com/opentracing-contrib/java-kafka-client/blob/31ce5260279c2cadf3d69c0acbd50f024afe4660/opentracing-kafka-client/src/main/java/io/opentracing/contrib/kafka/TracingKafkaUtils.java#L43

In OpenTelemetry, the default interceptor has the `onSend` calling this `buildAndInjectSpan` which gets the current context this way:

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/cd528e98fbbaa9bf6e48351d908d43e28a6e0073/instrumentation/kafka/kafka-clients/kafka-clients-2.6/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/KafkaTelemetry.java#L84

It's different from the OpenTracing one and record headers are not in the picture AFAICS.

Not an OpenTelemetry expert but it smells something is missing here for having it working out of box without any hack.

Reply:
> record headers are not in the picture

They are used / read here:
* https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/cd528e98fbbaa9bf6e48351d908d43e28a6e0073/instrumentation/kafka/kafka-clients/kafka-clients-2.6/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/KafkaTelemetry.java#L93

Reply:
I spent a quite of time debugging both OpenTracing and OpenTelemetry internals.
So you are right that OpenTracing is heavily based on using the Kafka record headers through the `uber-trace-id`. Thanks to it to be set in the Vert.x event loop thread (where the HTTP send span is created), the interceptor in a different Vert.x worker thread is able to get the parent span from that `uber-trace-id` working out of box.
With OpenTelemetry, the interceptor doesn't work like this and the context which makes sense in the Vert.x event loop thread (where the HTTP send span is created) doesn't exist in the interceptor (in the Vert.x worker thread) because everything OTel uses is a ThreadLocal storage for context (which of course isn't shared across threads :-)).
I am going to start a discussion in the OTel community because it means that the Kafka support doesn't work out of box and always needs an hack (as you are currently done). I was wondering why they didn't use the same OpenTracing way.

Reply:
I just opened a issue/discussion in the open-telemetry java related project https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/6297

Reply:
@ppatierno this Kafka OTel interceptor support is actually my implementation :-), but I did it by the guidelines of existing integration implementations (via that Instrumenter API). And that API, as I mentioned, needs an explicit current context / span.
The propagation or/aka storage can be changed - as ContextStorage is an abstraction, but I didn't find any other useful impl, where ThreadLocal is the default one.

Reply:
@alesj at this point the question is how the other implementations deal with cross-thread context ... I don't think that Kafka is the only use case. The current implementation is not that useable in the end because any user would need your hack by creating a custom interceptor class, handling an map of spans and so on.

Reply:
Yeah, good question -- any async impl (not just Kafka, etc) would have problems with this.
I guess they manage to pass-on the span somehow, w/o that map hack, etc ... ?

Reply:
I guess we need to find a way to wrap Vertx's executor in this Context::wrap .. any idea where this could be done?